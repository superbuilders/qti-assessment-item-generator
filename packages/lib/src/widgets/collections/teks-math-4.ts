import type { WidgetCollection } from "@/widgets/collections/types"
import { allWidgetDefinitions } from "@/widgets/definitions"

export const teksMath4Collection: WidgetCollection<
	Pick<
		typeof allWidgetDefinitions,
		| "protractorAngleDiagram"
		| "nPolygon"
		| "dotPlot"
		| "pictograph"
		| "tapeDiagram"
		| "numberLineWithAction"
		| "fractionModelDiagram"
		| "quantityFractionalDiagram"
		| "symmetryDiagram"
		| "stemLeafPlot"
		| "dataTable"
		| "geometricPrimitiveDiagram"
		| "decoratedNumber"
	>,
	readonly [
		"protractorAngleDiagram",
		"nPolygon",
		"dotPlot",
		"pictograph",
		"tapeDiagram",
		"numberLineWithAction",
		"fractionModelDiagram",
		"quantityFractionalDiagram",
		"symmetryDiagram",
		"stemLeafPlot",
		"dataTable",
		"geometricPrimitiveDiagram",
		"decoratedNumber"
	]
> = {
	name: "teks-math-4",
	widgets: {
		protractorAngleDiagram: allWidgetDefinitions.protractorAngleDiagram,
		nPolygon: allWidgetDefinitions.nPolygon,
		dotPlot: allWidgetDefinitions.dotPlot,
		pictograph: allWidgetDefinitions.pictograph,
		tapeDiagram: allWidgetDefinitions.tapeDiagram,
		numberLineWithAction: allWidgetDefinitions.numberLineWithAction,
		fractionModelDiagram: allWidgetDefinitions.fractionModelDiagram,
		quantityFractionalDiagram: allWidgetDefinitions.quantityFractionalDiagram,
		symmetryDiagram: allWidgetDefinitions.symmetryDiagram,
		stemLeafPlot: allWidgetDefinitions.stemLeafPlot,
		dataTable: allWidgetDefinitions.dataTable,
		geometricPrimitiveDiagram: allWidgetDefinitions.geometricPrimitiveDiagram,
		decoratedNumber: allWidgetDefinitions.decoratedNumber
	},
	widgetTypeKeys: [
		"protractorAngleDiagram",
		"nPolygon",
		"dotPlot",
		"pictograph",
		"tapeDiagram",
		"numberLineWithAction",
		"fractionModelDiagram",
		"quantityFractionalDiagram",
		"symmetryDiagram",
		"stemLeafPlot",
		"dataTable",
		"geometricPrimitiveDiagram",
		"decoratedNumber"
	]
} as const
