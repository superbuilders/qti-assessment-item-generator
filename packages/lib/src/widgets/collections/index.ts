import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
