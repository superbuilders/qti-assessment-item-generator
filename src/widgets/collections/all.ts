import type { WidgetCollection } from "@/widgets/collections/types"
import { allWidgetDefinitions } from "@/widgets/definitions"

function toReadonlyKeys<T extends Record<string, unknown>>(
	obj: T
): ReadonlyArray<keyof T & string> {
	const keys = Object.keys(obj)
	// Narrow via a runtime check that keys are strings
	return keys.filter((k): k is keyof T & string => typeof k === "string")
}

export const allWidgetsCollection: WidgetCollection<
	typeof allWidgetDefinitions,
	ReadonlyArray<keyof typeof allWidgetDefinitions & string>
> = {
	name: "all",
	widgets: allWidgetDefinitions,
	widgetTypeKeys: toReadonlyKeys(allWidgetDefinitions)
} as const
