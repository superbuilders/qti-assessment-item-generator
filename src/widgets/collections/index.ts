import { allWidgetsCollection } from "./all"
import { fourthGradeMathCollection } from "./fourth-grade-math"
import { mathCoreCollection } from "./math-core"
import { scienceCollection } from "./science"
import { simpleVisualCollection } from "./simple-visual"
import { teksMath4Collection } from "./teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "./types"
