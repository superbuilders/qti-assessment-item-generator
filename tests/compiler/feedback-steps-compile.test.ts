import { describe, expect, test } from "bun:test"
import { compile } from "@/compiler/compiler"
import type { AssessmentItemInput } from "@/core/item"
import { allWidgetsCollection } from "@/widgets/collections/all"

describe("Compiler: feedback-as-steps", () => {
	test("should render preamble + steps with correct colors and structure", async () => {
		const item: AssessmentItemInput<[]> = {
			identifier: "feedback-steps-item",
			title: "Feedback Steps Support",
			responseDeclarations: [
				{
					identifier: "RESPONSE",
					cardinality: "single",
					baseType: "identifier",
					correct: "A"
				}
			],
			body: [
				{ type: "paragraph", content: [{ type: "text", content: "What is 2 + 2?" }] },
				{ type: "interactionRef", interactionId: "ci" }
			],
			widgets: null,
			interactions: {
				ci: {
					type: "choiceInteraction",
					responseIdentifier: "RESPONSE",
					prompt: [{ type: "text", content: "Choose:" }],
					choices: [
						{
							identifier: "A",
							content: [{ type: "paragraph", content: [{ type: "text", content: "4" }] }]
						},
						{
							identifier: "B",
							content: [{ type: "paragraph", content: [{ type: "text", content: "5" }] }]
						}
					],
					shuffle: true,
					minChoices: 1,
					maxChoices: 1
				}
			},
			feedbackPlan: {
				mode: "combo",
				dimensions: [{ responseIdentifier: "RESPONSE", kind: "binary" }],
				combinations: [
					{
						id: "FB__RESPONSE_CORRECT",
						path: [{ responseIdentifier: "RESPONSE", key: "CORRECT" }]
					},
					{
						id: "FB__RESPONSE_INCORRECT",
						path: [{ responseIdentifier: "RESPONSE", key: "INCORRECT" }]
					}
				]
			},
			feedback: {
				FEEDBACK__OVERALL: {
					RESPONSE: {
						CORRECT: {
							content: {
								preamble: {
									correctness: "correct",
									summary: [{ type: "text", content: "You identified the sum correctly." }]
								},
								steps: [
									{
										type: "step",
										title: [{ type: "text", content: "Understand Addition" }],
										content: [
											{
												type: "paragraph",
												content: [
													{ type: "text", content: "2 + 2 means combining two groups of 2." }
												]
											}
										]
									},
									{
										type: "step",
										title: [{ type: "text", content: "Confirm the Result" }],
										content: [
											{
												type: "paragraph",
												content: [
													{ type: "text", content: "The total is " },
													{ type: "math", mathml: "<mn>4</mn>" },
													{ type: "text", content: "." }
												]
											}
										]
									}
								]
							}
						},
						INCORRECT: {
							content: {
								preamble: {
									correctness: "incorrect",
									summary: [{ type: "text", content: "The sum of 2 + 2 is not 5." }]
								},
								steps: [
									{
										type: "step",
										title: [{ type: "text", content: "Review Addition" }],
										content: [
											{
												type: "paragraph",
												content: [
													{ type: "text", content: "When adding, count all items together." }
												]
											}
										]
									},
									{
										type: "step",
										title: [{ type: "text", content: "Find the Correct Answer" }],
										content: [
											{
												type: "paragraph",
												content: [
													{ type: "text", content: "2 + 2 = " },
													{ type: "math", mathml: "<mn>4</mn>" }
												]
											}
										]
									}
								]
							}
						}
					}
				}
			}
		}
		const xml = await compile(item, allWidgetsCollection)

		// Preamble checks
		expect(xml).toContain("qti-feedback-preamble")
		expect(xml).toContain("Correct! Fantastic work.")
		expect(xml).toContain("You identified the sum correctly")
		expect(xml).toContain("Not quite! Try again.")
		expect(xml).toContain("The sum of 2 + 2 is not 5")

		// Steps checks
		expect(xml).toContain("qti-steps")
		expect(xml).toContain("qti-step")
		expect(xml).toContain("qti-step-index")
		expect(xml).toContain("qti-step-title")
		expect(xml).toContain("Understand Addition")
		expect(xml).toContain("Review Addition")

		// Color palette checks (step 1 blue, step 2 emerald)
		expect(xml).toContain("#2563EB") // blue
		expect(xml).toContain("#059669") // emerald

		expect(xml).toMatchSnapshot()
	})

	test("should enforce minimum 2 steps via schema validation", async () => {
		const itemWithOneStep: AssessmentItemInput<[]> = {
			identifier: "one-step-item",
			title: "One Step",
			responseDeclarations: [
				{
					identifier: "RESPONSE",
					cardinality: "single",
					baseType: "identifier",
					correct: "A"
				}
			],
			body: [
				{ type: "paragraph", content: [{ type: "text", content: "Question" }] },
				{ type: "interactionRef", interactionId: "ci" }
			],
			widgets: null,
			interactions: {
				CI: {
					type: "choiceInteraction",
					responseIdentifier: "RESPONSE",
					prompt: [],
					choices: [
						{ identifier: "A", content: [] },
						{ identifier: "B", content: [] }
					],
					shuffle: true,
					minChoices: 1,
					maxChoices: 1
				}
			},
			feedbackPlan: {
				mode: "fallback",
				dimensions: [{ responseIdentifier: "RESPONSE", kind: "binary" }],
				combinations: [
					{ id: "CORRECT", path: [] },
					{ id: "INCORRECT", path: [] }
				]
			},
			feedback: {
				FEEDBACK__OVERALL: {
					CORRECT: {
						content: {
							preamble: {
								correctness: "correct",
								summary: [{ type: "text", content: "Correct." }]
							},
							steps: [
								{
									type: "step",
									title: [{ type: "text", content: "Only Step" }],
									content: [{ type: "paragraph", content: [{ type: "text", content: "Done." }] }]
								}
							]
						}
					},
					INCORRECT: {
						content: {
							preamble: {
								correctness: "incorrect",
								summary: [{ type: "text", content: "Incorrect." }]
							},
							steps: [
								{
									type: "step",
									title: [{ type: "text", content: "Only Step" }],
									content: [
										{ type: "paragraph", content: [{ type: "text", content: "Try again." }] }
									]
								}
							]
						}
					}
				}
			}
		}

		// Should throw due to schema validation (steps.min(2))
		await expect(compile(itemWithOneStep, allWidgetsCollection)).rejects.toThrow()
	})
})
