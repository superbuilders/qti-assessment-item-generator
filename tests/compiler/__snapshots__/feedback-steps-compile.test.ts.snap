// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Compiler: feedback-as-steps should render preamble + steps with correct colors and structure 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<qti-assessment-item
    xmlns="http://www.imsglobal.org/xsd/imsqtiasi_v3p0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.imsglobal.org/xsd/imsqtiasi_v3p0 https://purl.imsglobal.org/spec/qti/v3p0/schema/xsd/imsqti_asiv3p0p1_v1p0.xsd http://www.w3.org/1998/Math/MathML https://purl.imsglobal.org/spec/mathml/v3p0/schema/xsd/mathml3.xsd"
    identifier="feedback-steps-item"
    title="Feedback Steps Support"
    time-dependent="false"
    xml:lang="en-US">
    <qti-response-declaration identifier="RESPONSE" cardinality="single" base-type="identifier">
        <qti-correct-response>
            <qti-value>A</qti-value>
        </qti-correct-response>
    </qti-response-declaration>
    <qti-outcome-declaration identifier="SCORE" cardinality="single" base-type="float">
        <qti-default-value><qti-value>0</qti-value></qti-default-value>
    </qti-outcome-declaration>
    <qti-outcome-declaration identifier="FEEDBACK__OVERALL" cardinality="single" base-type="identifier"/>
    <qti-item-body>
        <p>What is 2 + 2?</p>
        <div><qti-choice-interaction response-identifier="RESPONSE" shuffle="true" min-choices="1" max-choices="1">
            <qti-prompt>Choose:</qti-prompt>
            <qti-simple-choice identifier="A"><p>4</p></qti-simple-choice>
            <qti-simple-choice identifier="B"><p>5</p></qti-simple-choice>
        </qti-choice-interaction></div>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_CORRECT" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="correct" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#047857;">Correct! Fantastic work.</p>
      <p style="margin:8px 0 0 0; color:#111827;">You identified the sum correctly.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Understand Addition</p>
        <p>2 + 2 means combining two groups of 2.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Confirm the Result</p>
        <p>The total is <math xmlns="http://www.w3.org/1998/Math/MathML"><mn>4</mn></math>.</p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
        <qti-feedback-block outcome-identifier="FEEDBACK__OVERALL" identifier="FB__RESPONSE_INCORRECT" show-hide="show">
            <qti-content-body><div class="qti-feedback-preamble" data-correctness="incorrect" style="margin-bottom:16px;">
      <p style="margin:0; font-weight:700; color:#B91C1C;">Not quite! Try again.</p>
      <p style="margin:8px 0 0 0; color:#111827;">The sum of 2 + 2 is not 5.</p>
    </div><ol class="qti-steps" style="margin:16px 0; padding:0; list-style:none; counter-reset: qti-step;"><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #2563EB; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#2563EB; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">1</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Review Addition</p>
        <p>When adding, count all items together.</p>
      </li><li class="qti-step" style="position:relative; margin:16px 0; padding:16px 16px 16px 56px; border:1px solid #E5E7EB; border-left:6px solid #059669; border-radius:12px; background:#F9FAFB; box-shadow:0 1px 2px rgba(0,0,0,0.04);">
        <span class="qti-step-index" style="position:absolute; left:16px; top:16px; width:28px; height:28px; border-radius:9999px; background:#059669; color:#FFFFFF; display:inline-flex; align-items:center; justify-content:center; font-weight:700; font-size:14px;">2</span>
        <p class="qti-step-title" style="margin:0 0 8px 0; font-weight:700; font-size:16px; color:#111827;">Find the Correct Answer</p>
        <p>2 + 2 = <math xmlns="http://www.w3.org/1998/Math/MathML"><mn>4</mn></math></p>
      </li></ol></qti-content-body>
        </qti-feedback-block>
    </qti-item-body>
    <qti-response-processing>
        
    <qti-response-condition>
        <qti-response-if>
            <qti-match>
                <qti-variable identifier="RESPONSE"/>
                <qti-correct identifier="RESPONSE"/>
            </qti-match>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_CORRECT</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else>
            <qti-set-outcome-value identifier="FEEDBACK__OVERALL">
                <qti-base-value base-type="identifier">FB__RESPONSE_INCORRECT</qti-base-value>
            </qti-set-outcome-value>
        </qti-response-else>
    </qti-response-condition>
    <qti-response-condition>
        <qti-response-if>
            <qti-and>
                <qti-match><qti-variable identifier="RESPONSE"/><qti-correct identifier="RESPONSE"/></qti-match>
            </qti-and>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">1.0</qti-base-value></qti-set-outcome-value>
        </qti-response-if>
        <qti-response-else>
            <qti-set-outcome-value identifier="SCORE"><qti-base-value base-type="float">0.0</qti-base-value></qti-set-outcome-value>
        </qti-response-else>
    </qti-response-condition>
    </qti-response-processing>
</qti-assessment-item>"
`;
